api/src/main/java/org/openmrs/module/oauth2/api/smart/db/hibernate/SmartAppDAO.java

```diff
@@ -0,0 +1,47 @@
+package org.openmrs.module.oauth2.api.smart.db.hibernate;
+
+import java.util.List;
+
+import org.hibernate.Query;
+import org.hibernate.Session;
+import org.openmrs.User;
+import org.openmrs.module.oauth2.api.smart.model.SmartApp;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SmartAppDAO extends HibernateSmartDAO<SmartApp> {
+
+	/**
+	 * You must call this before using any of the data access methods, since it's not actually
+	 * possible to write them all with compile-time class information due to use of Generics.
+	 *
+	 * @param mappedClass
+	 */
+	protected SmartAppDAO(Class<SmartApp> mappedClass) {
+		super(mappedClass);
+	}
+
+	private SmartAppDAO() {
+		super(SmartApp.class);
+	}
+
+	public List<SmartApp> getAllSmartAppsForClientDeveloper(User clientDeveloper) {
+		String queryString = "from org.openmrs.module.oauth2.api.smart.model.SmartApp where creator.userId = :client_developer_id";
```

Comment:
does model SmartApp.java contain the column "creator"? Are you sure this query is correct?

Reply:
You caught me !! 

Copied from here... I didn't understand that query though.. :p

https://github.com/openmrs/openmrs-module-oauth2/blob/a9e6b796fa0cb4a40f1d6d3a40a201dee01ef0df/api/src/main/java/org/openmrs/module/oauth2/api/db/hibernate/ClientDAO.java#L34

Reply:
I mean i have understood the concepts here.. but.. this particular line confused me :)

Reply:
Look, the Client.java class (OAuth2 client) extends BaseOpenmrsData (which contains the fields created_by, modified_by, etc.) as you can see here (https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/BaseOpenmrsData.java). But your SMART application neither extends BaseOpenmrsData or OAuth Client class for that matter. Hence, you can't use that param here. This was a fundamental fault I thought I'd wait for you to find out until you start writing tests (as they would have failed and upon debugging you may have found this issue). So, in order to use ANYTHING related to an OAuth2 client your SmartApp class needs to extend "Client" class (Oauth2 class). You missed this in the previous pull requests. I thought I'd wait to see if you correct it while writing DAOs but you didn't :P So from next time, make sure you understand what's what and don't miss out on these small details. Otherwise, you'll have a hard time debugging other major errors that'll come in the project

Reply:
What I want you to do now, is to leave this PR as it is. Complete impl and DAO classes (with proper understanding :P) and then write tests to see if your api layer is ready.

Reply:
I hope you understand me. If not, we can do a call today evening or tomorrow :)

Reply:
To properly understand this, what you can do is see the classes, see what properties they inherit and from where. Also, look at the columns of "oauth2_client" :)

Reply:
What happened was... First i thought of extending the Client class itself.. as i thought each smart app is a client... But by doing so.. tests were failing... Application context failed to load.... Then i also though i must extend the  BaseOpenmrsData class... But doing so by just looking at other classes.... I didnt had a clue why we need to extend that class... So i dropped the idea...

Reply:
I actually haven't tested whatevr code i wrote... Thats why i think.. i must test my code so that... I could get errors.. and that may lead to let me understand what i have missed..

Reply:
That's weird

Reply:
I also thought of checking the logs but the logs simply said application context failed and that too for all the beans...

Reply:
Did they fail only after you extended the classes or you made some other changes as well?

Reply:
I will show you that on our call.. i am comfortable with any time.... I think i should add impl methods too.. then we can have a call before i start to test the api layer..

Reply:
Only after extending... As soon as i removed extends Client everything was fine

Reply:
Try if you can resolve those errors, if not, we have a simple turnaround from this. We already have ClientDAO and the columns in "oauth2_client" don't need to be in "oauth2_smart_app", so essentially when you want a smart apps from a client developer, you want oauth clients from that client developer that are actually smart apps. SO make use of ClientDAO here :)

Reply:
And frame your query in that way :) Either you resolve the errors you get while extending (which I am sure pretty sure has to do with the different scheme of columns). "OAuth Client" extends BaseOMRSdata but at the same time, also makes the columns in the DB. That isn't the case for "SmatApp". So think of modifying your DAO logic. In short, there is NO need to make a DAO method to get SmartApp using clientDeveloper's id. We can implement this in our "impl" classes and not in DAO classes. SmartDAO should only give you what it's supposed to get (i.e. mapped portion of SmartApp class)